### PR TITLE
Switch build system to Meson

### PR DESCRIPTION
**Pull request by ueno**
This adds a basic support for building and installing the provider
with Meson.  To test:

```console
meson setup build
meson compile -C build
meson test -C build --num-processes 1
```

Note that parallel execution of tests are currently not working, as
the same port number is used in multiple tests.

Here is some benchmark:
```console
$ time sh -c 'autoreconf -fi && ./configure && make && make check'
18.18s user 2.97s system 73% cpu 28.639 total

$ time sh -c 'meson setup build && meson compile -C build && meson test -C build --num-processes 1'
10.61s user 1.82s system 81% cpu 15.173 total
```

### CODE REVIEW

**Comment by simo5** [comment]
ok so I think a switch to meson is something we can consider, but it would have to be a full switch, definitely not going to maintain the same stuff in parallel in two places

FYI ignore debian, macos test failures for now they got the broken NSS 3.94;
but they should only fail tests and nothing else of course

Anyway for a change like this we'd have to wait until debian and macos upgrade to 3.95 in a month because we definitely need to ensure all the platforms work end-to-end

**Comment by ueno** [comment]
@simo5 I think this is almost ready, though I would need to take a closer look at the current CI failures (macOS build, address sanitizer with Debian, scan-build, and check style).

After the xz backdoor, it seems to be more relevant to switch to a simpler build system that doesn't allow extensions in a implicit/obfuscated way.

**Comment by simo5** [comment]
Given you are removing the old build system, can you provide a amke file with targets that correspond to today high level targets that just turn around and run meson commands?

At least:
- make
- make check
- make check-stile
- make check-style-show
- make check-style-fix
- make generate-code
- make generate-docs

File: packaging/pkcs11-provider.spec
Comment (simo5) [inline packaging/pkcs11-provider.spec:64]: does this print logs on failure like the previous line did?

File: .github/workflows/scan-build.yml
Comment (simo5) [inline .github/workflows/scan-build.yml:25]: Why does this need to be spun out as an additional build script?

File: .github/workflows/scan-build.yml
Comment (simo5) [inline .github/workflows/scan-build.yml:31]: why change from a simple file to a directory?

File: .github/workflows/build.yml
Comment (simo5) [inline .github/workflows/build.yml:61]: I really dislike that meson forces you to create a build directory, however if it is necessary can we name it "builddir" as is customary (afaik) with meson projects?
Why prefix it with "_" anyway ?

**Review by simo5 — CHANGES_REQUESTED** [review]
Mostly looks good, just some nitpicks

File: .github/workflows/style.yml
Comment (simo5) [inline .github/workflows/style.yml:17]: I think that either automake or libtool may have been implicitly be pulling in a c compiler, but meson does not, so you'll have to explicitly now.

File: .github/workflows/scan-build.yml
Comment (ueno) [reply .github/workflows/scan-build.yml:25]: This is the standard way of passing options to scan-build with Meson. See https://mesonbuild.com/howtox.html#use-clang-static-analyzer.

File: .github/workflows/build.yml
Comment (ueno) [reply .github/workflows/build.yml:61]: That is because some projects have `build` directly already. I can change it to `builddir`.

File: .github/workflows/scan-build.yml
Comment (ueno) [reply .github/workflows/scan-build.yml:31]: Actually, it hasn't changed: the logs of scan-build are always stored in a directory, afaik.

File: .github/workflows/style.yml
Comment (ueno) [reply .github/workflows/style.yml:17]: But here, `$COMPILER` is specified; where does it come from?

File: .github/workflows/scan-build.yml
Comment (simo5) [inline .github/workflows/scan-build.yml:16]: sounds like this $COMPILER variable is some spurious remnant, probably froma a copy of an old build.yml that was used to jump-start the scan-build action.
I think we can drop it, and/or, if needed, replace it with clang, but I guess clang-analyzer in the next line already pulls in all the required deps.

File: .github/workflows/style.yml
Comment (simo5) [reply .github/workflows/style.yml:17]: old remnant, but given we use clang to check style, perhaps we should just install clang instead of gcc as the compiler here.

File: .reuse/dep5
Comment (simo5) [inline .reuse/dep5:19]: should we drop provider.exports here ?
Is it still used once we move to provider.map ?

File: Makefile.am
Comment (simo5) [inline Makefile.am:40]: Where are the replacements for check-style, check-style-show, check-stuyle-fix, generate-code and generate-docs ?

**Review by simo5 — CHANGES_REQUESTED** [review]
Current changes look good,
however there is loss of functionality in the removal of the base Makefile.am that needs to be resolved before this PR can be adopted.

File: .github/workflows/integration.yml
Comment (simo5) [inline .github/workflows/integration.yml:25]: this backslash is incorrect

File: .github/workflows/integration.yml
Comment (simo5) [inline .github/workflows/integration.yml:31]: Is there a way to set some variable so that -C builddir does not have to be repeated on each command ?

**Comment by Jakuje** [comment]
> Uhm I wonder why the integration tests haven't run ...

Likely because of the backslash.

File: .reuse/dep5
Comment (ueno) [reply .reuse/dep5]: Removed.

File: Makefile.am
Comment (ueno) [reply Makefile.am:40]: I think we could add a Makefile for those "maintainer-mode" targets. Let me try in the next iteration.

File: .reuse/dep5
Comment (ueno) [reply .reuse/dep5]: Removed.

File: Makefile.am
Comment (ueno) [reply Makefile.am:40]: Done.

File: .github/workflows/address-sanitizer.yml
Comment (Jakuje) [inline .github/workflows/address-sanitizer.yml:31]: This is not a note for the meson change, but generic note to the way how the github actions are used here now.

When we were moving the CI of OpenSC from something previous to GH actions, I wanted to keep the bare minimum in the GH yml files and keep most of the logic in separate shell scripts, such as:

https://github.com/OpenSC/OpenSC/blob/master/.github/setup-linux.sh

This avoids when we change some dependencies to do it in half a dozen of yaml files.

This does not have to be changed as this PR, but I wanted to mention it here if you would find it useful, we can do it later.

File: meson.build
Comment (simo5) [inline meson.build:60]: If the correct thing is to call these .so, why are we changing them to .dylib here?

File: .reuse/dep5
Comment (simo5) [inline .reuse/dep5:11]: Why GNUmakefile instead of the common Makefile?

**Review by simo5 — COMMENTED** [review]
Just a couple nits/comments?

File: .reuse/dep5
Comment (ueno) [reply .reuse/dep5:11]: There are a couple of minor motivations here. First, you wouldn't need to care about compatibility with other make implementations, as GNUmakefile is only recognized by GNU make. Secondly, with GNU make, you could write the current rules slightly simpler, using wildcards without shell functions (I included this change in https://github.com/latchset/pkcs11-provider/pull/304/commits/30c77f5acda9f11c143a4c37504ddb35809eb195 but can revert it if you don't like it).

File: meson.build
Comment (ueno) [reply meson.build:60]: @neverpanic could you elaborate this?

File: .reuse/dep5
Comment (simo5) [reply .reuse/dep5:11]: I think we are making it hard enough for other platforms already with the switch to meson, let's just use a standard Makefile please.

Comment (ueno) [reply .reuse/dep5]: Reverted.

File: .github/workflows/integration.yml
Comment (ueno) [reply .github/workflows/integration.yml:30]: Not that I know of, I'm afraid.

File: tests/test-wrapper
Comment (simo5) [inline tests/test-wrapper:51]: Hi Daiki,
do not worry about this, I already tested a correct solution (this one is broken, already been there) and will opena  PR for after we merge yours.

**Review by simo5 — CHANGES_REQUESTED** [review]

File: src/signature.c
Comment (ueno) [reply src/signature.c:882]: Done.

File: tests/test-wrapper
Comment (ueno) [reply tests/test-wrapper]: OK, I removed this commit.

**Review by simo5 — CHANGES_REQUESTED** [review]
Ok I think we are really close.
The bind test failure is a regression in bind and we will waive it.

However I just realized that this PR will break the coverity scan.
Coverity relies on the command "make" to generate a build, it may even pass variables there.

Unfortunately it cannot be easily tested on a PR as the coverity scan can only run against an action already merged in the main tree.

But I can attempt a test in a local checkout to confirm that just providing a make :all target that call meson compile will work

**Comment by simo5** [comment]
So I think coverity depends only on a PATH change, injecting a different gcc in a coverity/bin path or similar.

How will meson interact with that ?

When I run `meson compile -vC builddir` it gives me what looks like a pseudo-command line as it identifies the compiler as just 'cc', so iut is not clear to me if meson executes the compiler via an abslute path or will allow interposition by calling the compiler with it's relative name.

**Comment by simo5** [comment]
I did find his post though:
https://community.synopsys.com/s/question/0D5Hr00006VdMFsKAN/in-ninja-build-environment-covbuild-cannot-emit-c-files

Which means you may need to tweak the .github/workflows/coverity-scan.yml steps before we can merge this PR, sorry for not realizing this earlier.

**Comment by ueno** [comment]
> I did find his post though: https://community.synopsys.com/s/question/0D5Hr00006VdMFsKAN/in-ninja-build-environment-covbuild-cannot-emit-c-files
> 
> Which means you may need to tweak the .github/workflows/coverity-scan.yml steps before we can merge this PR, sorry for not realizing this earlier.

I added `env.CC: gcc` in the setup step, following this [example](https://github.com/OpenRC/openrc/blob/3d30b6fddaf92c612deea88d2ced7114ed1fcf9c/.github/workflows/coverity.yml#L19) linked from the coverity-scan-action [README](https://github.com/vapier/coverity-scan-action/blob/main/README.md#live-examples). Hopefully it should work.

**Comment by simo5** [comment]
@dueno we sill need a Makefile all target to call meson compile as the covscan build just calls `make` with no arguments

**Comment by ueno** [comment]
> @dueno we sill need a Makefile all target to call meson compile as the covscan build just calls `make` with no arguments

I can add such make rule, but doesn't coverity-scan-action take care of that already, given the openrc [example](https://github.com/OpenRC/openrc/blob/HEAD/.github/workflows/coverity.yml) mentioned above doesn't have Makefile at all?

**Comment by simo5** [comment]
I am going to ignore the MacOS failure as it seems a homebrew screwup, as all of this was working until today.
Hopefully it will get fixed soon, and if not we'll handle it separately anyway.

**Review by simo5 — APPROVED** [review]

**Comment by neverpanic** [comment]
The macOS CI failure is a Homebrew screw-up, see https://github.com/Homebrew/homebrew-core/issues/169728.

I can locally reproduce:
```
$ podman pull ghcr.io/homebrew/core/openssl/3:3.3.0@sha256:636c1743241379d29fb964fcccce1ea86f46fa66a85319c1ac59c9698e4ac809  # this is the pre-compiled OpenSSL for Homebrew
$ podman image save 1f09999b3994 | tar xO f35bb52aadbdae43149bf5eb3a3a4948c1149879d13d2041e0139cbd80a08535.tar | tar tv | grep libcrypto.dylib
lrwxrwxrwx  0 0      0           0 Apr  9 14:12 openssl@3/3.3.0/lib/libcrypto.dylib -> libcrypto.3.dylib
$ podman image save 1f09999b3994 | tar xO f35bb52aadbdae43149bf5eb3a3a4948c1149879d13d2041e0139cbd80a08535.tar | tar xO openssl@3/3.3.0/lib/pkgconfig/libcrypto.pc | head -1
libdir=@@HOMEBREW_CELLAR@@/openssl@3/3.3.0
```

^ there's a `/lib` missing at the end of that path in the pkg-config file.

**Pull request merged**
